src/pip/_internal/utils/unpacking.py

```diff
@@ -0,0 +1,286 @@
+# mypy: strict-optional=False
+
+from __future__ import absolute_import
+
+import logging
+import mimetypes
+import os
+import shutil
+import stat
+import tarfile
+import zipfile
+
+from pip._internal.exceptions import InstallationError
+from pip._internal.utils.misc import (
+    current_umask,
+    ensure_dir,
+    file_contents,
+    has_leading_dir,
+    is_svn_page,
+    rmtree,
+    split_leading_dir,
+)
+from pip._internal.utils.typing import MYPY_CHECK_RUNNING
+
+if MYPY_CHECK_RUNNING:
+    from typing import Optional
+    from pip._internal.models.link import Link
+    from pip._internal.utils.hashes import Hashes
+
+logger = logging.getLogger(__name__)
+
+WHEEL_EXTENSION = '.whl'
+BZ2_EXTENSIONS = ('.tar.bz2', '.tbz')
+XZ_EXTENSIONS = ('.tar.xz', '.txz', '.tlz', '.tar.lz', '.tar.lzma')
+ZIP_EXTENSIONS = ('.zip', WHEEL_EXTENSION)
+TAR_EXTENSIONS = ('.tar.gz', '.tgz', '.tar')
+ARCHIVE_EXTENSIONS = (
+    ZIP_EXTENSIONS + BZ2_EXTENSIONS + TAR_EXTENSIONS + XZ_EXTENSIONS)
+SUPPORTED_EXTENSIONS = ZIP_EXTENSIONS + TAR_EXTENSIONS
+
+try:
+    import bz2  # noqa
+    SUPPORTED_EXTENSIONS += BZ2_EXTENSIONS
+except ImportError:
+    logger.debug('bz2 module is not available')
+
+try:
+    # Only for Python 3.3+
+    import lzma  # noqa
+    SUPPORTED_EXTENSIONS += XZ_EXTENSIONS
+except ImportError:
+    logger.debug('lzma module is not available')
+
+
+def unpack_file_url(
+    link,  # type: Link
+    location,  # type: str
+    download_dir=None,  # type: Optional[str]
+    hashes=None  # type: Optional[Hashes]
+):
+    # type: (...) -> None
+    """Unpack link into location.
+
+    If download_dir is provided and link points to a file, make a copy
+    of the link file inside download_dir.
+    """
+    from pip._internal.download import (
+        is_dir_url,
+        url_to_path,
+        _check_download_dir,
+        _copy_file
```

Comment:
Also, this module shouldn't have dependencies on `download.py` (avoid introducing new nested imports, by the way). Rather, the functions in `download.py` that `unpack_file_url()` depends on should also be moved into this module. We don't want to have dependencies from `utils` to higher-level modules outside of it like `download.py`.